src/System.Net.Http/src/System/Net/Http/Managed/HttpConnectionPools.cs

```diff
@@ -31,14 +31,25 @@ public HttpConnectionPools(int maxConnectionsPerServer)
         {
             _maxConnectionsPerServer = maxConnectionsPerServer;
             _pools = new ConcurrentDictionary<HttpConnectionKey, HttpConnectionPool>();
-            _cleaningTimer = new Timer(s => ((HttpConnectionPools)s).RemoveStalePools(), this, CleanPoolTimeoutMilliseconds, CleanPoolTimeoutMilliseconds);
+            // Start out with the timer not running, since we have no pools.
+            _cleaningTimer = new Timer(s => ((HttpConnectionPools)s).RemoveStalePools(), this, Timeout.Infinite, Timeout.Infinite);
         }
 
         /// <summary>Gets a pool for the specified endpoint, adding one if none existed.</summary>
         /// <param name="key">The endpoint for the pool.</param>
         /// <returns>The retrieved pool.</returns>
-        public HttpConnectionPool GetOrAddPool(HttpConnectionKey key) =>
-            _pools.GetOrAdd(key, (k, s) => new HttpConnectionPool(s), _maxConnectionsPerServer);
+        public HttpConnectionPool GetOrAddPool(HttpConnectionKey key)
+        {
+            // If we currently don't have any pools, a new one will be added for this
+            // connection. We should then start reaping stale pools.
+            if (_pools.IsEmpty)
```

Comment:
Ah, good to know. Hm, it seems we fall into this same trap just about any way we try to access the dictionary. The only other alternative I can think is to cache whether or not the dictionary was empty. In `RemoveStalePools` the `IsEmpty` check could remain since it is running on a thread pool, and under lock we can update a cached value that it is empty. In the `GetOrAddPool` method, we could similarly check the cached value under lock in order to start the timer. There would still be a potential race here in that we could start the timer assuming that it will subsequently be non empty and the `RemoveStalePools` method could fire before we actually add the new pool. So we would have to add the new pool first and then start the timer. Thoughts? Can we afford the added size of the cache field?